dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/vitals/FrameRateVitalReader.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.vitals
+
+import android.view.Choreographer
+import java.util.concurrent.TimeUnit
+
+/**
+ * Reads the UI framerate based on the [Choreographer.FrameCallback].
+ */
+internal class FrameRateVitalReader : VitalReader, Choreographer.FrameCallback {
+
+    private var lastFrameTimestampNs: Long = 0L
+    private var lastFrameRate: Double = Double.NaN
+
+    // region VitalReader
+
+    override fun readVitalData(): Double? {
+        return if (lastFrameRate.isNaN()) null else lastFrameRate
+    }
+
+    // endregion
+
+    // region Choreographer.FrameCallback
+
+    override fun doFrame(frameTimeNanos: Long) {
+        if (lastFrameTimestampNs != 0L) {
+            val durationNs = (frameTimeNanos - lastFrameTimestampNs).toDouble()
+            if (durationNs > 0.0) {
+                lastFrameRate = ONE_SECOND_NS / durationNs
+            }
+        }
+        lastFrameTimestampNs = frameTimeNanos
+        Choreographer.getInstance().postFrameCallback(this)
+    }
```

Comment:
as far as i understand `readVitalData()` is called [every 1 second](https://github.com/DataDog/dd-sdk-android/blob/xgouchet/RUMM-1277%2Ftrack_refresh_rate/dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/RumFeature.kt#L162) from a `ScheduledThreadPoolExecutor`.

but `doFrame` method overwrites `lastFrameRate` data at every new frame.
`readVitalData()` doesn't return the real frame rate, it returns the latest frame rate when it's called. we will catch a UI jank **only if** it happened at the latest frame before `ScheduledThreadPoolExecutor` is fired.

am i missing something?

Reply:
I think @buranmert is right....we are only going to catch the last janky frame with this approach as we are always computing the `lastFrameRate` based on the last frame duration. I think an idea would be to compute a mean of these frame rates which will span from the moment the last `readVitalData` was called to now (the current moment the `readVitalData` was called)